### DISCUSSION TITLE
Swiftbar and compiled plugins

### DISCUSSION THREAD

**Discussion by bilal-fazlani** [swiftbar/SwiftBar#73] [General]
Hi, right now there are two ways to feed information to swifbar

First one is by writing a program that prints out menu items. 

Second one is by having some text inside program. For example

```
# <bitbar.title>Title goes here</bitbar.title>
```

This tells swiftbar about metadata of the plugin. For example title, version, etc.

Another example is 

```
# <swiftbar.hideAbout>true</swiftbar.hideAbout>
```

Here, by having this text, I am configuring the plugin to hide “about” menu item.

Coming to the main point of discussion,

Having text inside plugin works only for script files. Compiled plugins are binary files and it won’t be possible for swifbar to read those lines. 

I propose we decide an approach where compiled plugins can also talk to swiftbar. 

One way to achieve this is by reading reading the same things, `swiftbar.hideAbout`, `title` `version` etc from the output. In the same format i.e. 

```
# <swiftbar.hideAbout>true</swiftbar.hideAbout>
```

A plug in compiled with java can then just print it in the beginning. 

This will open up swiftbar to every programming language like java, c#, kotlin, scala, F#, etc.

**Comment by bilal-fazlani** [comment]
The format could differ. May be metadata can be a section at the bottom or top separated by some special string such as `===` or something else. 

Example

```

My menu | color=red
---
Other item | href=https://google.com
Something | bash=Blabla.sh refresh=true
===
author=bilal
hideAbout=true
```

**Comment by melonamin** [comment]
I have another idea I think worth exploring, we can try to use extended file attributes. Each property can be a separate attribute, or we an package it like [this](https://stackoverflow.com/a/8555633)

**Reply by bilal-fazlani** [reply]
In the plist approach, do we end up with two files - one plugin executable and another plist file ?

**Reply by bilal-fazlani** [reply]
How do we merge the files into one? 🤔

**Reply by melonamin** [reply]
You can add custom metadata to any file  on macOS with `xattr` command, think about it like EXIF info in photos.

I can prototype this later this week to see if it works as expected

**Reply by bilal-fazlani** [reply]
I tried xattr. It worked. 

It felt suitable for metadata like about url, author, version etc. 

Did not feel right for configuring plugin 

For example: hideAbout, hideTerminal etc... 

What are your thoughts ?

Also, btw, is plist approach related to xattr ?

**Reply by melonamin** [reply]
Sorry, haven't time to look into this yet.

> Also, btw, is plist approach related to xattr ?

- Create plist containing all the configuration properties 
- Convert it to a binary form
- Attach it as a single attribute to the binary file using xattr

You can check the example on SO from the link above

**Reply by bilal-fazlani** [reply]
Ok i will check again tonight

**Reply by bilal-fazlani** [reply]
Ok, this can work

**Comment by bilal-fazlani** [comment]
Any new thoughts for this? 😁

**Comment by bilal-fazlani** [comment]
@melonamin 

These are the steps I followed

I have a simple sh file. `hello.5s.sh`

```sh
#!/bin/bash

echo hi
```

and I have created an `info.plist`

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>NSCalendarsUsageDescription</key>
	<string>This app needs access to your calendar.</string>
	<key>NSPhotoLibraryUsageDescription</key>
	<string>This app needs access to your photos.</string>
</dict>
</plist>
```

I then converted the plist file to a binary format with the following command

```bash
plutil -convert binary1 -o infofile info.plist
```

this gives a a new file `infofile`

I now write the entire file as an attribute 

```bash
xattr -wx "com.bilal-fazlani:META" (xxd -ps infofile | tr -d \\n) hello.5s.sh
```

Note: the `tr -d \\n` is required to remove line breaks

**Comment by bilal-fazlani** [comment]
This command returned me the binary data
```bash
xattr -pl com.bilal-fazlani:META hello.5s.sh | xxd -r -p
```

But I am unable to read it back

```bash
xattr -pl com.bilal-fazlani:META hello.5s.sh | xxd -r -p | plutil -convert xml1 - -o -
```

**Comment by melonamin** [comment]
Hey, thanks  for experimenting with it, on a second thought this what I've come with.

Take metadata as it is in non binary plugin(in this case I've created `metadata.txt` file):
```
# <bitbar.title>Dark Mode Toggle</bitbar.title>
# <bitbar.version>v1.0</bitbar.version>
# <bitbar.author>huhu</bitbar.author>
# <bitbar.author.github>Phlooo</bitbar.author.github>
# <bitbar.desc>Simple toggle for macOS Dark Mode</bitbar.desc>
# <bitbar.image>https://i.imgur.com/khIIfMg.png</bitbar.image>
# <bitbar.dependencies>bash</bitbar.dependencies>
# <bitbar.abouturl>https://github.com/Phlooo/</bitbar.abouturl>
# <swiftbar.hideAbout>true</swiftbar.hideAbout>
# <swiftbar.hideRunInTerminal>true</swiftbar.hideRunInTerminal>
# <swiftbar.hideLastUpdated>true</swiftbar.hideLastUpdated>
# <swiftbar.hideDisablePlugin>true</swiftbar.hideDisablePlugin>
# <swiftbar.hideSwiftBar>true</swiftbar.hideSwiftBar>
```

Write it to an extended file attribute as base64:
`xattr -w "com.ameba.SwiftBar" "$(cat metadata.txt | base64)" <plugin_file>`

This is much more easier for the end user and will require less code on SwiftBar end.

**Reply by melonamin** [reply]
I you're OK with this approach I'll go ahead and implement it #105

**Reply by bilal-fazlani** [reply]
Yes this approach is simpler than binary plist approach. Let's go ahead with this approach 👍

**Reply by melonamin** [reply]
There is a beta build if you want to play with it https://github.com/swiftbar/SwiftBar/releases/tag/v1.2.0-beta-1

**Reply by bilal-fazlani** [reply]
I will give this a try

**Reply by bilal-fazlani** [reply]
Can I upgrade to beta version using brew?

**Reply by melonamin** [reply]
No, unfortunately not :(
